modules/3d/include/opencv2/3d/volume.hpp

```diff
@@ -5,124 +5,131 @@
 #ifndef OPENCV_3D_VOLUME_HPP
 #define OPENCV_3D_VOLUME_HPP
 
+#include "volume_settings.hpp"
+
 #include "opencv2/core/affine.hpp"
 
 namespace cv
 {
+
 class CV_EXPORTS_W Volume
 {
-   public:
-    Volume(float _voxelSize, Matx44f _pose, float _raycastStepFactor) :
-        voxelSize(_voxelSize),
-        voxelSizeInv(1.0f / voxelSize),
-        pose(_pose),
-        raycastStepFactor(_raycastStepFactor)
-    { }
-
-    virtual ~Volume(){};
-
-    CV_WRAP
-    virtual void integrate(InputArray _depth, float depthFactor, const Matx44f& cameraPose,
-                           const Matx33f& intrinsics, const int frameId = 0)                   = 0;
-    virtual void integrate(InputArray _depth, InputArray _rgb, float depthFactor,
-                           const Matx44f& cameraPose, const Matx33f& intrinsics,
-                           const Matx33f& rgb_intrinsics, const int frameId = 0)               = 0;
-    CV_WRAP
-    virtual void raycast(const Matx44f& cameraPose, const Matx33f& intrinsics,
-                         const Size& frameSize, OutputArray points, OutputArray normals) const = 0;
-    virtual void raycast(const Matx44f& cameraPose, const Matx33f& intrinsics, const Size& frameSize,
-                         OutputArray points, OutputArray normals, OutputArray colors) const    = 0;
-    CV_WRAP
-    virtual void fetchNormals(InputArray points, OutputArray _normals) const                   = 0;
-    CV_WRAP
-    virtual void fetchPointsNormals(OutputArray points, OutputArray normals) const             = 0;
-    CV_WRAP
-    virtual void reset()                                                                       = 0;
-
-    // Works for hash-based volumes only, otherwise returns 1
-    virtual int getVisibleBlocks(int /*currFrameId*/, int /*frameThreshold*/) const { return 1; }
-    virtual size_t getTotalVolumeUnits() const { return 1; }
-
-   public:
-    const float voxelSize;
-    const float voxelSizeInv;
-    const Affine3f pose;
-    const float raycastStepFactor;
-};
+public:
+    /** @brief Constructor of default volume - TSDF.
+    */
+    Volume();
+    /** @brief Constructor of custom volume.
+    * @param vtype the volume type [TSDF, HashTSDF, ColorTSDF].
+    * @param settings the custom settings for volume.
+    */
+    Volume(VolumeType vtype, const VolumeSettings& settings);
+    ~Volume();
 
-struct CV_EXPORTS_W VolumeParams
-{
-    enum VolumeKind
-    {
-        TSDF        = 0,
-        HASHTSDF    = 1,
-        COLOREDTSDF = 2
-    };
-
-    /** @brief Kind of Volume
-        Values can be TSDF (single volume) or HASHTSDF (hashtable of volume units)
+    /** @brief Integrates the input data to the volume.
+
+    Camera intrinsics are taken from volume settings structure.
+
+    * @param frame the object from which to take depth and image data.
+    * @param pose the pose of camera in global coordinates.
+    */
+    void integrate(const OdometryFrame& frame, InputArray pose);
```

Comment:
Please also add to `integrate` docs:
For color TSDF a depth data should be registered with color data, i.e. have the same intrinsics & camera pose. This can be done using function `registerDepth()` from 3d module.